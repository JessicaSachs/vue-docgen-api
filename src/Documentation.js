class Documentation {
  constructor(initDocumentation = {}) {
    this._props = new Map(adaptToKeyValue(initDocumentation.props))
    this._data = new Map(adaptToKeyValue(initDocumentation.data))
  }

  set(key, value) {
    this._data.set(key, value)
  }

  get(key) {
    return this._data.get(key)
  }

  getPropDescriptor(propName) {
    var propDescriptor = this._props.get(propName)
    if (!propDescriptor) {
      this._props.set(propName, (propDescriptor = {}))
    }
    return propDescriptor
  }

  toObject() {
    var obj = {}

    for (var [key, value] of this._data) {
      obj[key] = value
    }

    if (this._props.size > 0) {
      obj.props = {}
      for (var [name, descriptor] of this._props) {
        obj.props[name] = descriptor
      }
    }

    return obj
  }
}

function adaptToKeyValue(obj) {
  return obj ? Object.keys(obj).map(k => [k, obj[k]]) : []
}

module.exports = Documentation
