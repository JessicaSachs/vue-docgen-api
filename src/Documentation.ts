import Map from 'ts-map'

export type BlockTag = ParamTag | Tag

export interface ParamType {
  name: string
  elements?: ParamType[]
}

export interface UnnamedParam {
  type?: ParamType
  description?: string | boolean
}

export interface Param extends UnnamedParam {
  name?: string
}

interface RootTag {
  title: string
}

export interface Tag extends RootTag {
  content: string | boolean
}

export interface ParamTag extends RootTag, Param {}

export interface DocBlockTags {
  description: string
  tags?: Array<ParamTag | Tag>
}

interface EventType {
  names: string[]
}

interface EventProperty {
  type: EventType
  name?: string
  description?: string | boolean
}

export interface EventDescriptor extends DocBlockTags {
  type?: EventType
  properties: EventProperty[] | undefined
}

export interface PropDescriptor {
  type?: { name: string; func?: boolean }
  description: string
  required?: string | boolean
  defaultValue?: { value: string; func?: boolean }
  tags: { [title: string]: BlockTag[] }
}

export interface MethodDescriptor {
  name: string
  description: string
  returns?: UnnamedParam
  tags?: { [key: string]: BlockTag[] }
  params?: Param[]
  modifiers: string[]
  [key: string]: any
}

export interface SlotDescriptor {
  description?: string
  bindings?: Record<string, any>
}

export interface ComponentDoc {
  displayName: string
  props: { [propName: string]: PropDescriptor } | undefined
  methods: MethodDescriptor[]
  slots: { [name: string]: SlotDescriptor }
  events?: { [name: string]: EventDescriptor }
  [key: string]: any
}

export class Documentation {
  private propsMap: Map<string, PropDescriptor>
  private methodsMap: Map<string, MethodDescriptor>
  private slotsMap: Map<string, SlotDescriptor>
  private eventsMap: Map<string, any>
  private dataMap: Map<string, any>
  constructor(initDocumentation?: ComponentDoc) {
    this.propsMap = new Map(
      initDocumentation ? adaptToKeyValue(initDocumentation.props) : undefined,
    )
    this.methodsMap = new Map(
      initDocumentation ? adaptToKeyValue(initDocumentation.methods) : undefined,
    )
    this.slotsMap = new Map(
      initDocumentation ? adaptToKeyValue(initDocumentation.slots) : undefined,
    )
    this.eventsMap = new Map(
      initDocumentation ? adaptToKeyValue(initDocumentation.events) : undefined,
    )
    this.dataMap = new Map(
      adaptToKeyValue(initDocumentation ? adaptToKeyValue(initDocumentation) : undefined),
    )
  }

  public set(key: string, value: any) {
    this.dataMap.set(key, value)
  }

  public get(key: string): any {
    return this.dataMap.get(key)
  }

  public getPropDescriptor(propName: string): PropDescriptor {
    return this.getDescriptor(propName, this.propsMap, () => ({
      description: '',
      tags: {},
    }))
  }

  public getMethodDescriptor(methodName: string): MethodDescriptor {
    return this.getDescriptor(methodName, this.methodsMap, () => ({
      name: methodName,
      modifiers: [],
      description: '',
      tags: {},
    }))
  }

  public getEventDescriptor(eventName: string): EventDescriptor {
    return this.getDescriptor(eventName, this.eventsMap, () => ({
      properties: [],
      description: '',
      tags: [],
    }))
  }

  public getSlotDescriptor(slotName: string): SlotDescriptor {
    return this.getDescriptor(slotName, this.slotsMap, () => ({
      description: '',
      tags: {},
    }))
  }

  public toObject(): ComponentDoc {
    const props = this.getPropsObject()
    const methods = this.getMethodsObject()
    const events = this.getEventsObject()
    const slots = this.getSlotsObject()

    const obj: { [key: string]: any } = {}
    this.dataMap.forEach((value, key) => {
      if (key) {
        obj[key] = value
      }
    })

    return {
      // initialize non null params
      description: '',
      tags: {},

      // set all the component params (override init values)
      ...obj,

      // set all the static properties
      displayName: obj.displayName,
      props,
      events,
      methods,
      slots,
    }
  }

  private getDescriptor<T>(name: string, map: Map<string, T>, init: () => T): T {
    let descriptor = map.get(name)
    if (!descriptor) {
      map.set(name, (descriptor = init()))
    }
    return descriptor
  }

  private getObjectFromDescriptor<T>(map: Map<string, T>): { [name: string]: T } | undefined {
    if (map.size > 0) {
      const obj: { [name: string]: T } = {}
      map.forEach((descriptor, name) => {
        if (name && descriptor) {
          obj[name] = descriptor
        }
      })
      return obj
    } else {
      return undefined
    }
  }

  private getPropsObject(): { [propName: string]: PropDescriptor } | undefined {
    return this.getObjectFromDescriptor(this.propsMap)
  }

  private getMethodsObject(): MethodDescriptor[] {
    const methods: MethodDescriptor[] = []
    this.methodsMap.forEach((descriptor, name) => {
      if (name && methods && descriptor) {
        methods.push(descriptor)
      }
    })
    return methods
  }

  private getEventsObject(): { [eventName: string]: EventDescriptor } | undefined {
    return this.getObjectFromDescriptor(this.eventsMap)
  }

  private getSlotsObject(): { [slotName: string]: SlotDescriptor } {
    return this.getObjectFromDescriptor(this.slotsMap) || {}
  }
}

/**
 * Transforms an object into an array of tuples [key, value]
 * @param {Object} obj
 * @returns {Array<[key, value]>} the retransfromed array and [] if obj is null
 */
function adaptToKeyValue(obj: { [key: string]: any } | undefined): Array<[string, any]> {
  const keyValuePairs: Array<[string, any]> = obj
    ? Object.keys(obj).map((k: string) => {
        const kvp: [string, any] = [k, obj[k]]
        return kvp
      })
    : []
  return keyValuePairs
}
