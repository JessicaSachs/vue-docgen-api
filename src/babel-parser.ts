import { File as BabelFile } from '@babel/types';
import { parse, ParserOptions } from '@babel/parser';

const babelParserOptions: ParserOptions = {
  sourceType: 'module',
  strictMode: false,
  tokens: true,
  plugins: [
    'jsx',
    'flow',
    'estree',
    'doExpressions',
    'objectRestSpread',
    'classProperties',
    'classPrivateProperties',
    'classPrivateMethods',
    'exportDefaultFrom',
    'exportNamespaceFrom',
    'asyncGenerators',
    'functionBind',
    'functionSent',
    'dynamicImport',
    'numericSeparator',
    'optionalChaining',
    'importMeta',
    'bigInt',
    'optionalCatchBinding',
    'throwExpressions',
    'nullishCoalescingOperator',
  ],
};

export default function buildParse(): { parse: (src: string) => BabelFile } {
  return {
    parse(src: string): BabelFile {
      return parse(src, babelParserOptions);
    },
  };
}
