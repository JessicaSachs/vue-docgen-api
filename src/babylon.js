const parser = require('@babel/parser')

const babelParserOptions = {
  sourceType: 'module',
  strictMode: false,
  tokens: true,
  plugins: [
    'jsx',
    'flow',
    'estree',
    'doExpressions',
    'objectRestSpread',
    'classProperties',
    'classPrivateProperties',
    'classPrivateMethods',
    'exportDefaultFrom',
    'exportNamespaceFrom',
    'asyncGenerators',
    'functionBind',
    'functionSent',
    'dynamicImport',
    'numericSeparator',
    'optionalChaining',
    'importMeta',
    'bigInt',
    'optionalCatchBinding',
    'throwExpressions',
    ['pipelineOperator', { proposal: 'minimal' }],
    'nullishCoalescingOperator',
  ],
}

function buildOptions(options = {}) {
  const parserOptions = {
    ...babelParserOptions,
    plugins: [...babelParserOptions.plugins],
  }
  if (options.legacyDecorators) {
    parserOptions.plugins.push('decorators-legacy')
  } else {
    parserOptions.plugins.push([
      'decorators',
      { decoratorsBeforeExport: options.decoratorsBeforeExport || false },
    ])
  }

  return parserOptions
}

export default function buildParse(options) {
  const parserOptions = buildOptions(options)

  return {
    parse(src) {
      return parser.parse(src, parserOptions)
    },
  }
}
