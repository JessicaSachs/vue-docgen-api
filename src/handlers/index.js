import displayNameHandler from './displayNameHandler'
import methodHandler from './methodHandler'
import propHandler from './propHandler'

export default [displayNameHandler, methodHandler, propHandler]
