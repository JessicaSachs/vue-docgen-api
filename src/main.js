import fs from 'fs'
import path from 'path'
import * as utils from './utils'
import parseSource from './parse'

export { default as parseSource } from './parse'

export { utils }

export function parse(file) {
  const singleFileComponent = /\.vue/i.test(path.extname(file))
  const source = fs.readFileSync(file, {
    encoding: 'utf-8',
  })
  return parseSource(source, singleFileComponent, file)
}
