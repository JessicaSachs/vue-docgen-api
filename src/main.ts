import * as fs from 'fs'
import { ComponentDoc } from './Documentation'
import parseSource from './parse'
import * as utils from './utils'

export { default as parseSource } from './parse'

export { utils }

export { ComponentDoc }

export function parse(filePath: string): ComponentDoc {
  const source = fs.readFileSync(filePath, {
    encoding: 'utf-8',
  })
  return parseSource(source, filePath)
}
