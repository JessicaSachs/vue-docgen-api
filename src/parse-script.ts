import { ParserPlugin } from '@babel/parser'
import { NodePath } from '@babel/traverse'
import * as bt from '@babel/types'
import buildParser from './babel-parser'
import { ComponentDoc, Documentation } from './Documentation'
import getRequiredExtendsDocumentations from './utils/getRequiredExtendsDocumentations'
import getRequiredMixinDocumentations from './utils/getRequiredMixinDocumentations'
import resolveExportedComponent from './utils/resolveExportedComponent'

// tslint:disable-next-line:no-var-requires
const deepmerge = require('deepmerge')

const ERROR_MISSING_DEFINITION = 'No suitable component definition found'

export default function parseScript(
  source: string,
  handlers: Array<(doc: Documentation, componentDefinition: NodePath) => void>,
  options: { lang: 'ts' | 'js'; filePath: string },
): { doc: ComponentDoc; ast: bt.File } {
  const plugins: ParserPlugin[] = options.lang === 'ts' ? ['typescript'] : ['flow']
  const ast = buildParser({ plugins }).parse(source)
  if (!ast) {
    throw new Error(ERROR_MISSING_DEFINITION)
  }
  const componentDefinitions = resolveExportedComponent(ast)

  if (componentDefinitions.length === 0) {
    throw new Error(ERROR_MISSING_DEFINITION)
  }

  // extends management
  const extendsDocumentations =
    getRequiredExtendsDocumentations(ast, componentDefinitions, options.filePath) || {}

  // mixins management
  const mixinsDocumentations = getRequiredMixinDocumentations(
    ast,
    componentDefinitions,
    options.filePath,
  )

  // merge all the varnames found in the mixins
  const mixinDocs = Object.keys(mixinsDocumentations).reduce((acc, mixinVar) => {
    return deepmerge(acc, mixinsDocumentations[mixinVar])
  }, extendsDocumentations)

  const vueDocArray = executeHandlers(handlers, componentDefinitions, mixinDocs)
  return { doc: vueDocArray[0], ast }
}

function executeHandlers(
  localHandlers: Array<(doc: Documentation, componentDefinition: NodePath) => void>,
  componentDefinitions: NodePath[],
  mixinsDocumentations: { props?: any; data?: any },
) {
  return componentDefinitions.map((compDef) => {
    const documentation = new Documentation(mixinsDocumentations)
    localHandlers.forEach((handler) => handler(documentation, compDef))
    return documentation.toObject()
  })
}
