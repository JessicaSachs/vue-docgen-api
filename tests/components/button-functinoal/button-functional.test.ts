import * as path from 'path';
import { ComponentDoc } from '../../../src/Documentation';
import { parse } from '../../../src/main';

const button = path.join(__dirname, './MyButton.vue');
let docButton: ComponentDoc;

describe('tests button functional', () => {
  beforeEach((done) => {
    docButton = parse(button);
    done();
  });

  it('should have a slot.', () => {
    expect(Object.keys(docButton.slots).length).toEqual(1);
  });

  it('should have a default slot.', () => {
    expect(docButton.slots.default).not.toBeUndefined();
  });

  it('the default slot should have "Use this slot default" as description', () => {
    expect(docButton.slots.default.description).toEqual('Use this slot default');
  });

  it('should the component has size prop description equal The size of the button', () => {
    expect(docButton.props.size.description).toEqual('The size of the button');
  });

  it('should match the snapshot', () => {
    expect(docButton).toMatchSnapshot();
  });
});
