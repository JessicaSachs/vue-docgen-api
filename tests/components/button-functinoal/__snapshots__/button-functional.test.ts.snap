// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests button functional should match the snapshot 1`] = `
Object {
  "comment": "",
  "description": "",
  "displayName": "MyButton",
  "events": Object {},
  "methods": Array [],
  "props": Object {
    "size": Object {
      "comment": "",
      "defaultValue": Object {
        "func": false,
        "value": "\\"normal\\"",
      },
      "description": "The size of the button",
      "required": "",
      "tags": Object {},
      "type": Object {
        "name": "string",
      },
    },
  },
  "slots": Object {
    "default": Object {
      "description": "Use this slot default",
    },
  },
  "tags": Object {},
}
`;
